### PR TITLE
[TypeSpec Validation] Removing git diff rule

### LINKED ISSUE #25707: [TypeSpec Validation] Move "git diff" into "compile" and "format" rules
Currently, the "compile" and "format" rules run first, then the "git diff" rule checks for any diffs created by these steps in the whole repo.  This has a few problems:

1. When validating multiple specs on a dev machine, the diffs from one spec can "pollute" the results of other specs.
2. It prevents validating specs in parallel, since the diff is repo-wide (#25694)
3. It creates a dependency between rules, while independent rules would be preferable.
4. It groups the changes made by "compile" and "format" into a single diff.

I believe we can fix the above without any loss of functionality via the following:

1. Run "git diff" at the end of "compile" and "format"
2. Scope the "git diff" to the top-level spec folder (e.g. `specification/foo`).  Files may be modified anywhere under this folder, but neither compile nor format should modify files outside this folder.

### PR DESCRIPTION
**Pull request by ckairen**
closes #25707

### CODE REVIEW

File: eng/tools/typespec-validation/src/rules/format.ts [hunk 1]
```diff
@@ -10,9 +11,17 @@ export class FormatRule implements Rule {
     // Format parent folder to include shared files
 
     let [err, stdOutput, errorOutput] = await host.runCmd(`npx tsp format "../**/*.tsp"`, folder);
+
+    const gitDiffResult = await gitDiffTopSpecFolder(host, folder);
+    stdOutput += gitDiffResult.stdOutput;
+    if (!gitDiffResult.success) {
+      errorOutput += gitDiffResult.errorOutput;
+      errorOutput += `\nFiles has been gerenate/changed after tsp format, please ensure all files are included in the branch.`;
```
Comment (weshaggard) [inline eng/tools/typespec-validation/src/rules/format.ts:19]: We should instruct them to run tsp format to fix this.

File: eng/tools/typespec-validation/src/utils.ts [hunk 2]
```diff
@@ -33,3 +34,25 @@ export function normalizePath(folder: string, path: PlatformPath = defaultPath)
...
+  let success = true;
+  let errorOutput: string | undefined;
+
+  if (!gitStatus.isClean()) {
+    success = false;
+    errorOutput = `Files generated: ${gitStatus.not_added}\n`;
+    errorOutput += `Files modified: ${gitStatus.modified}`;
+  }
+
+  return {
```
Comment (weshaggard) [inline eng/tools/typespec-validation/src/utils.ts:53]: Do we also need to undo the changes somewhere?

File: eng/tools/typespec-validation/src/rules/compile.ts [hunk 3]
```diff
@@ -44,6 +45,14 @@ export class CompileRule implements Rule {
       errorOutput += stderr;
     }
 
+    const gitDiffResult = await gitDiffTopSpecFolder(host, folder);
```
Comment (mikeharder) [inline eng/tools/typespec-validation/src/rules/compile.ts:48]: Don't call "git diff" if "tsp compile" or "tsp format" fails.  The result will probably be misleading or useless.

File: eng/tools/typespec-validation/src/rules/format.ts [hunk 4]
```diff
@@ -10,9 +11,17 @@ export class FormatRule implements Rule {
     // Format parent folder to include shared files
 
     let [err, stdOutput, errorOutput] = await host.runCmd(`npx tsp format "../**/*.tsp"`, folder);
+
+    const gitDiffResult = await gitDiffTopSpecFolder(host, folder);
```
Comment (mikeharder) [inline eng/tools/typespec-validation/src/rules/format.ts:15]: Don't call "git diff" if "tsp compile" or "tsp format" fails. The result will probably be misleading or useless.

File: eng/tools/typespec-validation/src/rules/compile.ts
Comment (mikeharder) [inline eng/tools/typespec-validation/src/rules/compile.ts:1]: Please add tests simulating all the conditions in the new code you are adding (e.g. "tsp compile" fails, it succeeds but there is a git diff, and it suceeds but no git diff).

**Review by mikeharder — CHANGES_REQUESTED** [review]
Add tests and skip diff if previous command failed.

File: eng/tools/typespec-validation/src/utils.ts (hunk 2 above)
Comment (ckairen) [reply eng/tools/typespec-validation/src/utils.ts:53]: the undo script is here: https://github.com/Azure/azure-rest-api-specs/blob/main/eng/scripts/Validate-TypeSpec.ps1#L27-L28
we're refraining from undoing it because we don't want devs to accidentally delete their local dev work every time they run tsv

Comment (weshaggard) [reply eng/tools/typespec-validation/src/utils.ts:53]: If we are splitting these checks into individual git diffs it would seem that one failing check will cause others to also fail if we don't undo between the steps. Perhaps we need an option for tsv to undo as it goes or if it should leave modified files.

Comment (ckairen) [reply eng/tools/typespec-validation/src/utils.ts:53]: We are stopping on first occurrence of error. We think that additional rule execution will make the output more confusing. So I don't think we need to undo between steps

**Review by mikeharder — APPROVED** [review]
